src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/LanguageServices/EditorConfigFileWatcher.cs

```diff
@@ -0,0 +1,271 @@
+            SortedSet<string> updatedEditorConfigFilePaths = GetEditorConfigFilePaths(newTree, projectUpdate);
+
+            var filesToStopWatching = _filesBeingWatched.Keys.Where(p => !updatedEditorConfigFilePaths.Contains(p)).ToArray();
+            await UnregisterFileWatchersAsync(filesToStopWatching);
+
+            var filesToStartWatching = updatedEditorConfigFilePaths.Where(p => !_filesBeingWatched.TryGetValue(p, out var _)).ToArray();
+            await RegisterFileWatchersAsync(filesToStartWatching);
+        }
+
+        private async Task UnregisterFileWatchersAsync(string[] filesToStopWatching)
+        {
+            IVsAsyncFileChangeEx fileChangeService = await _fileChangeService.GetValueAsync();
+            foreach (string editorConfigFilePath in filesToStopWatching)
+            {
+                uint cookie = _filesBeingWatched[editorConfigFilePath];
+                if (cookie != VSConstants.VSCOOKIE_NIL)
+                {
+                    await fileChangeService.UnadviseFileChangeAsync(cookie);
+                }
+
+                _filesBeingWatched.Remove(editorConfigFilePath);
+            }
+        }
+
+        private async Task RegisterFileWatchersAsync(string[] filesToStartWatching)
+        {
+            IVsAsyncFileChangeEx fileChangeService = await _fileChangeService.GetValueAsync();
+            foreach (string editorConfigFilePath in filesToStartWatching)
+            {
+                _filesBeingWatched[editorConfigFilePath] =
+                    await fileChangeService.AdviseFileChangeAsync(
+                        editorConfigFilePath,
+                        _VSFILECHANGEFLAGS.VSFILECHG_Add | _VSFILECHANGEFLAGS.VSFILECHG_Del,
+                        sink: this);
+            }
+        }
+
+        /// <summary>
+        /// Initialize the watcher.
+        /// </summary>
+        protected override async Task InitializeCoreAsync(CancellationToken cancellationToken)
+        {
+            // Explicitly get back to the thread pool for the rest of this method so we don't tie up the UI thread;
+            await TaskScheduler.Default;
+
+            await _projectTasksService.LoadedProjectAsync(() =>
+                {
+                    _filesBeingWatched = new Dictionary<string, uint>(StringComparers.Paths);
+                    _taskDelaySchedulerCancellationTokenSource = CreateLinkedCancellationTokenSource();
+                    _taskDelayScheduler = new TaskDelayScheduler(
+                        s_notifyDelay,
+                        _projectServices.ThreadingService,
+                        _taskDelaySchedulerCancellationTokenSource.Token);
+
+                    // The tree source to get changes to the tree so that we can identify *when* the set of .editorconfig files changes.
+                    ProjectDataSources.SourceBlockAndLink<IProjectVersionedValue<IProjectTreeSnapshot>> treeSource = _fileSystemTreeProvider.Tree.SyncLinkOptions();
```

Comment:
Looking at usage here, it seems like you are using data flow just to grab the directory of the project?